src/TraverseTracks.cpp

```diff
@@ -0,0 +1,158 @@
+ */
+TraverseTracks::~TraverseTracks() {
+}
+
+
+/**
+ * @brief Loops over Tracks, applying the provided kernel to all segments and
+ *        the functionality described in onTrack(...) to all Tracks.
+ * @details The segment formation method imported from the TrackGenerator
+ *          during construction is used to redirect to the appropriate looping
+ *          scheme.
+ * @param kernel The MOCKernel to apply to all segments
+ */
+void TraverseTracks::loopOverTracks(MOCKernel* kernel) {
+  switch (_segment_formation) {
+    case EXPLICIT_2D:
+      loopOverTracks2D(kernel);
+      break;
+    default:
+      log_printf(ERROR, "Segment formation type not currently supported");
+  }
+}
+
+
+/**
+ * @brief Loops over Tracks by parallel groups to avoid potential conflicts,
+ *        applying the provided kernel to all segments and the functionality
+ *        described in onTrack(...) to all Tracks.
+ * @details The segment formation method imported from the TrackGenerator
+ *          during construction is used to redirect to the appropriate looping
+ *          scheme.
+ * @param kernel The MOCKernel to apply to all segments
+ */
+void TraverseTracks::loopOverTracksByParallelGroup(MOCKernel* kernel) {
+  switch (_segment_formation) {
+    case EXPLICIT_2D:
+      loopOverTracksByParallelGroup2D(kernel);
+      break;
+    default:
+      log_printf(ERROR, "Segment formation type not currently supported");
+  }
+}
+
+
+/**
+ * @brief Loops over all explicit 2D Tracks
+ * @details The onTrack(...) function is applied to all 2D Tracks and the
+ *          specified kernel is applied to all segments. If NULL is provided
+ *          for the kernel, only the onTrack(...) functionality is applied.
+ * @param kernel The MOCKernel to apply to all segments
+ */
+void TraverseTracks::loopOverTracks2D(MOCKernel* kernel) {
+
+  /* Loop over all parallel tracks for each azimuthal angle */
+  Track** tracks_2D = _track_generator->getTracks();
+  int num_azim = _track_generator->getNumAzim();
+  for (int a=0; a < num_azim/2; a++) {
+    int num_xy = _track_generator->getNumX(a) + _track_generator->getNumY(a);
+#pragma omp for
+    for (int i=0; i < num_xy; i++) {
+
+      Track* track_2D = &tracks_2D[a][i];
+
+      /* Apply the kernel to segments if necessary */
+      if (kernel != NULL) {
+        kernel->newTrack(track_2D);
+        traceSegmentsExplicit(track_2D, kernel);
+      }
+
+      /* Operate on the Track */
+      segment* segments = track_2D->getSegments();
+      onTrack(track_2D, segments);
+    }
+  }
+}
+
+
+/**
+ * @brief Loops over all explicit 2D Tracks by parallel group
+ * @details This function is the same as loopOverTracks2D except that
+ *          Tracks are traversed in a safe order such that there are no
+ *          conflicts in exchanging data between Tracks in parallel.
+ * @param kernel The MOCKernel to apply to all segments
+ */
+void TraverseTracks::loopOverTracksByParallelGroup2D(MOCKernel* kernel) {
```

Comment:
Do you want to just remove the parallel group and store the old track fluxes since that's what we are transitioning to for 3D?